src/seq/mod.rs

```diff
@@ -591,37 +615,64 @@ impl<T> SliceRandom for [T] {
     }
 
     fn partial_shuffle<R>(
         &mut self, rng: &mut R, amount: usize,
     ) -> (&mut [Self::Item], &mut [Self::Item])
-    where R: Rng + ?Sized {
-        // This applies Durstenfeld's algorithm for the
-        // [Fisher–Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm)
-        // for an unbiased permutation, but exits early after choosing `amount`
-        // elements.
-
-        let len = self.len();
-        let end = if amount >= len { 0 } else { len - amount };
+    where
+        R: Rng + ?Sized,
+    {
+        let m = self.len().saturating_sub(amount);
 
-        for i in (end..len).rev() {
-            // invariant: elements with index > i have been locked in place.
-            self.swap(i, gen_index(rng, i + 1));
+        // The algorithm below is based on Durstenfeld's algorithm for the
+        // [Fisher–Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm)
+        // for an unbiased permutation.
+        // It ensures that the last `amount` elements of the slice are randomly selected from the whole slice.
+
+        // The loop invariant is that elements in the range [m,i) are randomly selected from the first i elements.
+        // This is true initially because the range is empty as i = m.
+        // The loop body swaps the element at i with a random one of the first i + 1 elements, then increments i.
+        // After the swap, the element at i will be randomly selected from the first i + 1 elements.
+        // Because of the invariant, before the swap, every element in the range [m,i) was randomly selected from the first i elements.
+        // Each of those elements has a 1 in i chance of being swapped with the element at i.
+        // Therefore, after the swap, the elements in the range [m,i) will be randomly selected from the first i + 1 elements.
+        // But the element at i was also randomly selected from the first i + 1 elements.
+        // So the elements in the range [m,i+1) are all randomly selected from the first i + 1 elements.
+        // So before we increment i, the elements in the range [m,i + 1) are randomly selected from the first i + 1 elements.
+        // Therefore, the loop invariant holds.        
+        // When the loop exits, elements in the range [m,length] will be randomly selected from the whole slice.
+        
+        
+        //`IncreasingUniform::next_index()` is about twice as fast as `gen_index` but only works for 32 bit integers
+        //Therefore we must use the slow method if the slice is longer than that.
+        if self.len() < (u32::MAX as usize) {
+            let mut chooser = IncreasingUniform::new(rng, m as u32);
+            for i in m..self.len() {
+                let index = chooser.next_index();
+                self.swap(i, index);
+            }
+        } else {            
+            for i in m..self.len() {
```

Comment:
Ermm, I'm pretty sure I've got this right. The last element gets swapped to a random place in the list so it has a m/len probability of being in the first m elements. Earlier elements are more likely to be chosen initially but can get booted out by later ones.  The `test_shuffle` test is checking this and I've also tried similar tests with longer lists and more runs. 

The reason I don't reverse the iterator is because the `increasing_uniform` needs `i` to increase and a decreasing version would be more complicated.

Reply:
Okay. We previously reversed since this way the proof by induction is easier. But we can also prove this algorithm works.

First, lets *not* use `m = len - amount` since in the last PR we used `m = amount`. I'll continue to use `end = len - amount`.

Lets say we have a list `elts = [e0, e1, .., ei, ..]` of length `len`. Elements are "chosen" if they appear in `elts[end..len]` after the algorithm; additionally we need to show that this slice is fully shuffled.

Algorithm is:
```rust
for i in end..len {
    elts.swap(i, rng.sample_range(0..=i));
}
```

For any length, for `amount = 0` or `amount = 1`, this is clearly correct. We'll prove by induction, assuming that the algorithm is already proven correct for `amount-1` and `len-1` (so that `end` does not change and the algorithm only has one last swap to perform).

Thus, we assume:
-   For any elt `ei`, we have `P(ei in elts[0..end]) = end/(len-1)` [here we say nothing about element order]
-   For any elt `ei`, for any `k in end..(len-1)`, `P(elts[k] = ei) = (amount-1)/(len-1)` [fully shuffled]

We perform the last step of the algorithm: `let x = sample_range(0..=len); elts.swap(len-1, x);`. Now:
-   Any element in `elts[0..end]` is moved to `elts[len-1]` with probability `1/len`, thus for any elt `ei` except `e_last`, `P(ei in elts[0..end]) = end/(len-1) * (len-1)/len = end/len`
-   For any elt `ei` previously in `elts[end..len-1]`, the chance it is not moved is `(len-1)/len`, thus, for these `ei`, for any `k in end..(len-1)`, `P(elts[k] = ei) = (amount-1)/(len-1) * (len-1)/len = (amount-1)/len`
-   For any elt `ei` previously in `elts[end..len-1]`, `P(elts[len-1] = ei) = 1/len`
-   The previous two points together imply that for any `ei` previously in `elts[end..len-1]`, for any `k in end..len`, `P(elts[k] = ei) = (amount-1+1)/len = amount/len`
-   Element `e_last` may appear in any position with probability `1/len`

Thus each element has chance `amount/len` to appear in `ents[end..len]` and this slice is fully shuffled.